experimental/schoology/export-assignment-data.py

```diff
@@ -0,0 +1,269 @@
+        gl = {
+            gp.id: gp.title for gp in gp_list if gp.id in self.desired_grading_periods
+        }
+        self.grading_periods_lookup = gl
+
+    def _create_file_name(self, base_name):
+        if not os.path.exists(self.output_path):
+            os.mkdir(self.output_path)
+
+        file_name = f'{base_name}.{datetime.now().strftime("%Y-%m-%d-%H-%M-%S")}.csv'
+        return os.path.join(self.output_path, file_name)
+
+    def _to_csv(self, data, base_name):
+        df = pd.DataFrame(data=data)
+        df.to_csv(self._create_file_name(base_name), mode="w", index=False)
+
+    def _get_students(self):
+        # Pull data from Schoology
+        student_list = self.sc_client.get_students()
+
+        # Reshape for our desired output
+        def dto_to_student(student):
+            return {
+                "uid": student.uid,
+                "school_id": student.school_id,
+                "school_uid": student.school_uid,
+                "name_first": student.name_first,
+                "name_middle": student.name_middle,
+                "name_last": student.name_last,
+                "username": student.username,
+                "email_address": student.primary_email,
+            }
+
+        return [dto_to_student(s) for s in student_list]
+
+    def _get_sections(self):
+        # TODO: planning to build a fresh Schoology API package, at which time we
+        # can delete the following extension method
+        from schoolopy.models import Section
+
+        def get_sections_for_course(Self):
+            resource = "courses/{}/sections".format(Self.id)
+
+            # TODO: make sure this handles situation where there are no sections for a course
+            return [Section(raw) for raw in self.sc_client.get(resource)["section"]]
+
+        schoolopy.models.Course.get_sections = get_sections_for_course
+        del get_sections_for_course
+        # End extension
+
+        # Pull data from Schoology
+        course_list = self.sc_client.get_courses()
+
+        section_list = []
+        for c in course_list:
+            # Filter only sections for the grading periods we care about.
+            section_list += [
+                c
+                for c in c.get_sections()
+                for gp in self.grading_periods_lookup
+                if gp in c.grading_periods
+            ]
+
+        # Reshape for our desired output
+        def dto_to_section(section):
+            return {
+                "id": section.id,
+                "course_id": section.course_id,
+                "school_id": section.school_id,
+                "course_title": section.course_title,
+                "course_code": section.course_code,
+                "section_title": section.section_title,
+                "grading_period_ids": section.grading_periods,
+                "grading_period_names": [
+                    self.grading_periods_lookup[gp] for gp in section.grading_periods
+                ],

```

Comment:
A section can be in multiple grading periods, for example ["first six weeks", "fall semester", "2020-2021 school year"]. In Slack we discussed creating separate output records for each school year. I ran out of time to change my code; for now the list is flattened into a single column.